src/Dialect/ONNX/ONNXOps.cpp

```diff
@@ -2158,35 +2158,44 @@ LogicalResult ONNXUnsqueezeOp::inferShapes(
     return emitError("Input tensor not ranked");
 
   auto operandTy = data().getType().cast<RankedTensorType>();
-  int inRank = operandTy.getRank();
+  auto elementType = data().getType().cast<ShapedType>().getElementType();
+  int64_t inRank = operandTy.getRank();
 
   ArrayAttr axisAttrs = axesAttr();
-  SmallVector<int, 4> axes;
-  int outRank = 0;
-  if (axisAttrs) {
-    outRank = inRank + axisAttrs.getValue().size();
-    for (auto axisAttr : axisAttrs.getValue()) {
-      int axis = axisAttr.cast<IntegerAttr>().getInt();
-      axis = axis >= 0 ? axis : (outRank + axis);
-      // Valid range
-      assert(axis >= -outRank && axis <= outRank - 1);
-      if (std::find(axes.begin(), axes.end(), axis) == axes.end())
-        axes.emplace_back(axis);
-      else
-        return emitError("Duplicated axes");
-    }
-  } else
+  if (!axisAttrs)
     return emitError("Axes attribute is required");
 
-  SmallVector<int64_t, 4> dims;
-  for (int i = 0, j = 0; i < outRank || j < inRank; ++i) {
-    if (std::find(axes.begin(), axes.end(), i) != axes.end()) {
-      dims.emplace_back(1);
-    } else {
-      dims.emplace_back(operandTy.getShape()[j++]);
+  SmallVector<int64_t, 4> axes;
+  bool hasNegativeAxis = false;
```

Comment:
NIT: a better name for this bool is "updateAxis" as technically the axis should not be negative once the rank is added...

Reply:
Comment on changing the negative value of axes:
1. It is not part of shape inference
2. What benefit does the change bring about? Since new op might be generated along rewriting, we always have to handle the possible negative value
3. In future, the axes will be a tensor. The IndexExpr to handle the negative value has to be there anyway.
My conclusion: no need to make the change.

Reply:
@chentong319 The way we have handled it before was to have such processing in Onnx Shape helper, which is then used both for shape inference and lowering. That is where we can handle the cases where the axis is known only at runtime.

Since this is to fix a bug, I suggest that we prioritize speed and we can generalize later as we see fit.